mpc_shifting_knapsack/shifting_knapsack.mpc

```diff
@@ -0,0 +1,150 @@
+import math
+
+
+program.use_edabit(True)
+
+
+def show_matrix(m):
+    for i in range(len(m)):
+        print_ln("%s ", m[i].reveal())
+    print_ln("")
+
+
+def knapsack(weights, values, C):
+    '''
+    The recursive param is to choose if solving the problem using the iterative
+    version or the recursive one.
+    '''
+    N = len(weights)
+    lwC = math.ceil(math.log2(C + 1)) + 1
+    ret = Array(1, sint)
+    ret[0] = 0
+
+    dp = Matrix(2, C + 1, sint)
+    dp.assign_all(0)
+
+    @for_range(N)
+    def _(i):
+        @for_range(C + 1)
+        def _(j):
+            dp[i % 2][j] = dp[(i + 1) % 2][j]
+
+        cyclic_shift(dp[(i + 1) % 2], C + 1 - weights[i], lwC)
+
+        @for_range(C + 1)
+        def _(j):
+            opt2 = dp[(i + 1) % 2][j] + values[i]
+            should_mov = (j >= weights[i]) * (opt2 > dp[i % 2][j])
+
+            dp[i % 2][j] = should_mov.if_else(opt2, dp[i % 2][j])
+
+        increased = dp[i % 2][C] > ret[0]
+
+        ret[0] = increased.if_else(dp[i % 2][C], ret[0])
+
+    return ret[0]
+
+
+def bitwise_xor(a, b):
+    bits_a = a.bit_decompose()
+    bits_b = b.bit_decompose()
+    bits_xor = Array(len(bits_a), sint)
+
+    for i in range(len(bits_a)):
+        bits_xor[i] = bits_a[i].bit_xor(bits_b[i])
+
+    return sint.bit_compose(bits_xor)
+
+
+def bitwise_and(a, b):
+    bits_a = a.bit_decompose()
+    bits_b = b.bit_decompose()
+    bits_and = Array(len(bits_a), sint)
+
+    for i in range(len(bits_a)):
+        bits_and[i] = bits_a[i].bit_and(bits_b[i])
+
+    return sint.bit_compose(bits_and)
+
+
+def cyclic_shift(values, K, max_log_K):
+    K = sint(K) 
+
+    @for_range(max_log_K)
+    def _(i):
+        should_mov = bitwise_and(K, (cint(1 << (max_log_K - i - 1)))) != 0
```

Comment:
`max_log_K - i - 1` is used 3 times in this function. Maybe create a new variable to store for better clarity.